frontend/src/lib/k8s/api/v2/watching.ts

```diff
@@ -0,0 +1,159 @@
+
+/**
+ * Create new WebSocket connection to the backend
+ */
+export async function openWebSocket<T>(
+  url: string,
+  {
+    protocols: moreProtocols = [],
+    type = 'binary',
+    cluster = getCluster() ?? '',
+    onMessage,
+  }: {
+    /**
+     * Any additional protocols to include in WebSocket connection
+     */
+    protocols?: string | string[];
+    /**
+     *
+     */
+    type: 'json' | 'binary';
+    /**
+     * Cluster name
+     */
+    cluster?: string;
+    /**
+     * Message callback
+     */
+    onMessage: (data: T) => void;
+  }
+) {
+  const path = [url];
+  const protocols = ['base64.binary.k8s.io', ...(moreProtocols ?? [])];
+
+  const token = getToken(cluster);
+  if (token) {
+    const encodedToken = btoa(token).replace(/=/g, '');
+    protocols.push(`base64url.bearer.authorization.k8s.io.${encodedToken}`);
+  }
+
+  if (cluster) {
+    path.unshift('clusters', cluster);
+
+    try {
+      const kubeconfig = await findKubeconfigByClusterName(cluster);
+
+      if (kubeconfig !== null) {
+        const userID = getUserIdFromLocalStorage();
+        protocols.push(`base64url.headlamp.authorization.k8s.io.${userID}`);
+      }
+    } catch (error) {
+      console.error('Error while finding kubeconfig:', error);
+    }
+  }
+
+  const socket = new WebSocket(makeUrl([BASE_WS_URL, ...path], {}), protocols);
+  socket.binaryType = 'arraybuffer';
+  socket.addEventListener('message', (body: MessageEvent) => {
+    const data = type === 'json' ? JSON.parse(body.data) : body.data;
+    onMessage(data);
+  });
+  socket.addEventListener('error', error => {
+    console.error('WebSocket error:', error);
+  });
+
+  return socket;
+}
+
+/**
+ * Listen for changes in a List of Kubernetes resources
+ */
+export async function watchList<T extends KubeObjectInterface>({
+  endpoint,
+  resourceVersion,
+  cluster = getCluster() ?? '',
+  queryParams,
+  onListUpdate,
+  itemClass,
+}: {
+  endpoint: KubeObjectEndpoint;
+  /**
+   * Pass a version of the list that you already have
+   *
+   * https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions
+   */
+  resourceVersion: string;
+  /** Cluster name */
+  cluster?: string;
+  queryParams?: Record<string, string>;
+  /**
+   * Change callback. Update function will apply the changes to the list
+   */
+  onListUpdate: (update: (oldList: KubeList<T>) => KubeList<T>) => void;
+  /** Class to instantiate items in the list */
+  itemClass: KubeObjectClass;
+}) {
+  const watchUrl = makeUrl([KubeObjectEndpoint.toUrl(endpoint)], {
+    ...queryParams,
+    watch: '1',
+    resourceVersion,
+  });
+
+  function update(event: KubeListUpdateEvent<T>) {
+    onListUpdate(list => KubeList.applyUpdate(list, event, itemClass));
+  }
+
+  const socket = await openWebSocket<KubeListUpdateEvent<T>>(watchUrl, {
+    type: 'json',
+    cluster,
+    onMessage: update,
+  });
+
+  return socket;
+}
+
+/**
+ * Listen for changes to a particular Kubernetes resource object
```

Comment:
Can you please add `@param` and `@returns` here?

Reply:
done